### DISCUSSION TITLE
I don't think the difficulty adjustment is reasonable

### DISCUSSION THREAD

**Discussion by coinsccg** [HarukaMa/aleo-pool-server#7] [Q&A]
pub async fn next_difficulty(&mut self) -> u64 {
        match self.current_difficulty.cmp(&self.next_difficulty) {
            cmp::Ordering::Less => {
                self.speed_1m.reset().await; // ??????
                self.current_difficulty = self.next_difficulty;
            }
            cmp::Ordering::Greater => {
                self.current_difficulty = ((0.9 * self.current_difficulty as f64) as u64).max(1); 
            }
            _ => {}
        }
        self.current_difficulty
    }

Why reset vecdeque?

**Comment by HarukaMa** [comment]
If the difficulty is just increased, there will be lower difficulty share still in the counter and it will affect the calculation. You can try removing that reset line and see what will happen after a difficulty increase.

That said, this was written before there are a good speedometer, so I’m planning to rewrite that part using proper miner speed tracking.

Closing the issue here, feel free to reopen if you have further questions.

**Comment by HarukaMa** [comment]
Oh, one more thing - there is a major rewrite going on with significant changes on database and prover protocol so I'd recommend not trying to deploy it right now.

**Comment by coinsccg** [comment]
> If the difficulty is just increased, there will be lower difficulty share still in the counter and it will affect the calculation. You can try removing that reset line and see what will happen after a difficulty increase.
> 
> That said, this was written before there are a good speedometer, so I’m planning to rewrite that part using proper miner speed tracking.
> 
> Closing the issue here, feel free to reopen if you have further questions.

When the difficulty value does not change, the speed will not change, but you also reset it

**Comment by HarukaMa** [comment]
`cmp::Ordering::Less`

**Comment by coinsccg** [comment]
`self.next_difficulty =
            ((self.current_difficulty as f64 * f64::sqrt(self.speed_1m.speed().await * 20f64)) as u64).max(1);`

As long as `self.speed_1m.speed().await * 20f64` is greater than 1 and remains unchanged, the value of `self.next_difficulty` will become larger and larger, will always meet this condition `self.current_difficulty.cmp(&self.next_difficulty)`

**Comment by HarukaMa** [comment]
full ctx for ref:
https://github.com/HarukaMa/aleo-pool-server/blob/0000034269db72c55009f0bd62b2b8cb27e9c24f/src/server.rs#L64-L88

The current difficulty target is to keep about 1 share per 20 seconds for each miner. So of course if `self.speed_1m.speed().await * 20f64` is greater than 1 the code would want to increase the next difficulty for that miner.

Also note that `speed_1m` is counted *differently* as it counts the *number* of shares submitted by the miner (share / sec), instead of miner's actual speed (proof / sec). I'm planning to change this in the next version.

> will always meet this condition `self.current_difficulty.cmp(&self.next_difficulty)`

Which arm are you referring to?
